pdfoutline.1.in
.\" -*- nroff -*-
.TH pdfoutline 1 "@DATE@" "@CMAKE_PROJECT_VERSION@" "@CMAKE_PROJECT_NAME@"
.SH NAME
pdfoutline \- add outlines (aka bookmarks) to PDF files
\" macros
.de SAMPLE
.br
.RS
.nf
.nh
..
.de ESAMPLE
.hy
.fi
.RE
..
.SH SYNOPSIS
.B pdfoutline
.I input.pdf outlines.txt output.pdf
.SH DESCRIPTION
\fBpdfoutline\fP reads input file given as first argument, adds outlines from text file given
as second argument, and saves result to file with name given as third argument.
.P
File with outlines information should consist of lines in the following format:
.SAMPLE
\fI<level>\fP \fI<page>\fP Outline text
.ESAMPLE
.P
\fI<level>\fP and \fI<page>\fP should be integers.
Each field should be separated by exactly one space symbol.
All values for \fI<level>\fP should be greater or equal than that of the first line.
Page numeration starts with 1.
.P
Outlines file can contain comments that start with # in first column.
Comments and empty lines are ignored.
The text is expected to be in UTF-8 encoding.
.SH OPTIONS
.B pdfoutline
accepts no options.
.SH EXAMPLES
Here is example of outlines data file:
.SAMPLE
0 1 Document title
1 1 Chapter 1
2 1 Chapter 1.1
2 2 Chapter 1.2
1 3 Chapter 2
.ESAMPLE
.P
Using this file will result in outlines like the following:
.SAMPLE
Document title
+-Chapter 1
| +-Chapter 1.1
| +-Chapter 1.2
+-Chapter 2
.ESAMPLE
.SH AUTHOR
.B pdfoutline
author is Eugeniy Meshcheryakov <eugen@debian.org>
.br
\fBpdfoutline\fP is part of \fBfntsample\fP and can be downoaded from
<@CMAKE_PROJECT_HOMEPAGE_URL@>.


ChangeLog
Changes in version 5.3
	* Support reproducible builds with PDF output (Khaled Hosny)

Changes in version 5.2
	* Fix handling of non-ASCII characters in pdfoutline

Changes in version 5.1
	* Make writing outlines with Cairo actually work
	* Fix typos

Changes in version 5.0
	* Add command line flag that allows to use pango for text layout
	  (by Khaled Hosny, requires pango >= 1.37)
	* Add possiblility to create PDF outline directly using cairo
	  (by Khaled Hosny, requires cairo >= 1.15.4)
	* Switch to CMake as build system
	* Add command line flag for loading Unicode blocks file during
	  runtime.

Changes in version 4.1:
	* Detect iconv and add LIBICONV to LDADD
	* Detect support for -Wl,--as-needed in configure
	* Those changes fix compilation on OS X

Changes in version 4.0:
	* Add a --no-embed option
	* Fix broken handling of font names

Changes in version 3.2:
	* Add support for SVG output (works correctly only for single page output)

Changes in version 3.1:
	* Add support for files that contain multiple fonts, like
	  TrueType Collections (.ttc)
	* Link with libm to avoid compilation failure with gold linker
	* Use silent-rules feature of automake 1.11

Changes in version 3.0:
	* Make it buildable with mawk again.
	* Switch to GPLv3.
	* Added localization support using gettext.
	* Added Ukrainian translation.
	* Added font scaling for fonts with large vertical metrics,
	  to better work with fonts like TibetanMachine

Changes in version 2.8:
	* It is now possible to use file Blocks.txt with dos line endings,
	  like in zipped Unicode 5.1 UCD.

Changes in version 2.7:
	* configure now checks for pangocairo >= 1.16, older pango does not
	  have pango_layout_get_line_readonly() required by fntsample
	* Fixed crash (segmentation fault) with some values for -i (and maybe
	  -x)

Changes in version 2.6:
	* pango is now used for drawing headers and numbers in font samples
	* Font sizes are now tunable using "--style" option. Default font
	  size for digits was made smaller for better look with fonts containing
	  glyphs outside BMP

Changes in version 2.5:
	* Added support for compiling with non-default fontconfig location
	* Added support for long options
	* Made some optimizations to text layout, this should make generated
	  PDFs smaller
	* Added options for selecting Unicode ranges to include in samples
	  (-i and -x)

Changes in version 2.4:
	* Added script for making pdf files with bookmarks (pdfoutline) and
	  option for fntsample to print bookmarks information (-l).
	  pdfoutline requires perl and PDF::API2 library

Changes in version 2.3:
	* Added "#include <unistd.h>" to get definition of getopt()

Changes in version 2.2:
	* Added support for fonts other than TrueType/OpenType
	* Improved errors handling


pdfoutline.pl
#! /usr/bin/env perl
# This file is in public domain
# Author: Eugeniy Meshcheryakov <eugen@debian.org>
#
# This program adds outlines to pdf files.
# Usage: pdfoutline input.pdf outline.txt out.pdf
#
# File given as second argument should contain outline information in
# form:
#
# <level> <page> Some text
#
# where <level> and <page> are integers. Values for <level> should be greater
# or equal than that of first line. Page numeration starts with 1.
#
# Outlines file can contain comments that start with # in first column. Comments
# and empty lines are ignored.
#
# Example file:
# 0 1 Document title
# 1 1 Chapter 1
# 2 1 Chapter 1.1
# 2 2 Chapter 1.2
# 1 3 Chapter 2
#
# This file will result in outline like the following:
#
# Document title
# +-Chapter 1
# | +-Chapter 1.1
# | +-Chapter 1.2
# +-Chapter 2

use strict;
use warnings;
use PDF::API2;
use Locale::TextDomain('@CMAKE_PROJECT_NAME@', '@CMAKE_INSTALL_FULL_LOCALEDIR@');
use POSIX qw(:locale_h);
use Encode qw(encode);

sub usage {
    printf __"Usage: %s input.pdf outline.txt out.pdf\n", $0;
}

# get first non-empty non-comment line
sub get_line {
    my ($F) = @_;
    my $line;

    while ($line = <$F>) {
        chomp $line;
        # skip comments ...
        next if $line =~ /^#/;
        # ... and empty lines
        next if $line eq q{};
        last;
    }
    return $line;
}

# Encode string to UTF-16BE with BOM if it contains non-ASCII characters
sub encode_pdf_text {
    my ($str) = @_;

    if ($str !~ /[^[:ascii:]]/) {
        return $str;
    } else {
        return encode('UTF-16', $str);
    }
}

sub add_outlines {
    my ($pdf, $parent, $line, $F) = @_;
    my $cur_outline;

    my ($level) = split / /, $line;

    MAINLOOP: while ($line) {
        my ($new_level, $page, $text) = split / /, $line, 3;

        if ($new_level > $level) {
            $line = add_outlines($pdf, $cur_outline, $line, $F);
            next MAINLOOP;
        } elsif ($new_level < $level) {
            return $line;
        } else {
            $cur_outline = $parent->outline;
            $cur_outline->title(encode_pdf_text($text));
            # FIXME it should be posible to make it easier
            my $pdfpage = $pdf->{pagestack}->[$page - 1];
            $cur_outline->dest($pdfpage);
        }

        $line = get_line($F);
    }
}

# Create new outlines object ignorig outlines that can be
# already present in the PDF file.
sub new_outlines {
    my ($pdf) = @_;

    require PDF::API2::Outlines;
    $pdf->{'pdf'}->{'Root'}->{'Outlines'} = PDF::API2::Outlines->new($pdf);
    my $obj = $pdf->{'pdf'}->{'Root'}->{'Outlines'};

    $pdf->{'pdf'}->new_obj($obj) unless $obj->is_obj($pdf->{'pdf'});
    $pdf->{'pdf'}->out_obj($obj);
    $pdf->{'pdf'}->out_obj($pdf->{'pdf'}->{'Root'});

    return $obj;
}

setlocale(LC_ALL, q{});

if ($#ARGV != 2) {
    usage;
    exit 1;
}

my ($inputfile, $outlinefile, $outputfile) = @ARGV;

my $pdf = PDF::API2->open($inputfile);

open my $outline_fh, '<:encoding(UTF-8)', $outlinefile
    or die __x("Cannot open outline file '{outlinefile}'",
               outlinefile => $outlinefile);

my $line = get_line($outline_fh);

# create new outlines here, don't try to use old ones
my $outlines = new_outlines($pdf);

add_outlines($pdf, $outlines, $line, $outline_fh) if $line;
close $outline_fh;

$pdf->saveas($outputfile);

exit 0;
